Add File System Support via pyodide.FS

fullzer4 opened the pull request:
# Add File System Support via pyodide.FS

This PR implements comprehensive file system operations in the PyodideSandbox, addressing **issue #34** (Allow attaching files) and **issue #26** (Preserve newlines when printing). 

## Key Features

### File System Operations
- **File Attachment**: Pre-load files into sandbox with `attach_file()` and `attach_files()`
- **Standard Python I/O**: Full support for `open()`, `Path()`, and standard file operations
- **Multiple File Types**: Handle text, binary files (CSV, images, etc.)
- **Directory Support**: Organize files in nested directory structures

### Bug Fixes
- **Resolves #26**: Fixed print output concatenation - `print(5); print(5)` now correctly outputs `5\n5` instead of `55`
- **Resolves #34**: Complete file attachment functionality implementation

## Implementation Details

- Added `FileSystemOperation` interface for defining file operations
- Implemented file system operations in `main.ts` using pyodide.FS
- Migrated from BaseTool to StructuredTool for dynamic descriptions

## Usage Example

```python
# pip install langgraph-codeact "langchain[anthropic]"
import asyncio
from langchain_sandbox import PyodideSandboxTool
from langgraph.prebuilt import create_react_agent

# Define sandbox tool with filesystem support
sandbox_tool = PyodideSandboxTool(
    enable_filesystem=True,
    allow_net=True,
)

# Attach sample data
sales_data = """...csv_data"""

sandbox_tool.attach_file("sales.csv", sales_data)

# Create agent with sandbox tool
agent = create_react_agent("anthropic:claude-3-7-sonnet-latest", [sandbox_tool])

query = """Please analyze the sales data and tell me:
1. What is the total revenue by category?
2. Which region has the highest sales?
3. What are the top 3 best-selling products by revenue?

Use pandas to read the CSV file and perform the analysis."""

async def run_agent(query: str):
    async for chunk in agent.astream({"messages": query}):
        print(chunk)

if __name__ == "__main__":
    asyncio.run(run_agent(query))
```

## References

- [Pyodide File System Documentation](https://pyodide.org/en/stable/usage/file-system.html)

fullzer4 committed:
feat(filesystem): add MEMFS filesystem support

fullzer4 committed:
Merge branch 'main' of github.com:fullzer4/langchain-sandbox into fullzer4/add_fyle_system

fullzer4 committed:
feat(sandbox): file attachment operations

fullzer4 committed:
fix: improve file system implementation and file handling

fullzer4 committed:
Merge pull request #1 from fullzer4/fullzer4/add_fyle_system

Fullzer4/add fyle system

fullzer4 committed:
fix: conventions

fullzer4 committed:
Merge pull request #2 from fullzer4/fullzer4/add_fyle_system

fix: conventions

fullzer4 committed:
fix: _build_command place convention

fullzer4 committed:
Merge pull request #3 from fullzer4/fullzer4/add_fyle_system

fix: _build_command place convention

fullzer4 committed:
fix: better filesystem operations and file handling

fullzer4 committed:
feat: implement dynamic filesystem support with flexible tool architecture

fullzer4 committed:
fix: follow python conventions

fullzer4 committed:
fix: typescript conventions

fullzer4 committed:
fix: rollback some tests

fullzer4 committed:
fix: permissions

fullzer4 committed:
feat: README changes

fullzer4 committed:
fix: unified PyodideSandboxTool with complete backward compatibility

fullzer4 committed:
fix: clean README

fullzer4 committed:
fix: lint

eyurtsev reviewed (commented):
@fullzer4 this PR is really amazing. Thank you for putting all of this together! I left a few questions about the extra entrypoints in the code and would like to figure out how to avoid them if possible.

Thanks also for fixing the various issues in the existing code! You caught more than one bug!

fullzer4 commented on README.md:
```diff
@@ -35,9 +34,6 @@ LangChain Sandbox provides a secure environment for executing untrusted Python c
 ## 💡 Example Usage
 
 
-> [!warning]
```
This was an unintentional change that I didn't notice during my edits. I'll revert this change

fullzer4 commented on libs/sandbox-py/langchain_sandbox/pyodide.py:
```diff
@@ -736,3 +1176,119 @@ async def _arun(
         return tool_result
+
+
+class PyodideSandboxStructuredTool:
```
You're absolutely right. The wrapper was created for convenience and discoverability, but it does add an unnecessary entrypoint. I'll remove the PyodideSandboxStructuredTool

fullzer4 committed:
fix: Remove PyodideSandboxStructuredTool and fix unintentional README changes

fullzer4 commented on libs/sandbox-py/langchain_sandbox/pyodide.py:
```diff
@@ -610,9 +847,212 @@ class PyodideSandboxToolInput(BaseModel):
+        self._sandbox.attach_binary_file(path, content)
+        self._sync_sandbox.attach_binary_file(path, content)
+        # Update both BaseTool and StructuredTool descriptions
+        new_description = self._build_description()
```
I can see this could break custom instructions like "only use jax for calculations."

I'm honestly a bit torn between two approaches:

**Option 1**: Keep auto-updating descriptions but add safeguards
- Detect when user provides custom description during init
- For custom descriptions: only append file info, preserve original content  
- For default descriptions: continue full auto-updates

**Option 2**: Remove auto-updating descriptions entirely
- Remove the automatic description updates from attach_file()
- Provide manual method: `tool.update_description()` for users who want it
- Simpler, more predictable behavior

What do you think works better here?

fullzer4 replied:
> @fullzer4 let me know what your thoughts are on this!
> 
> The interface could take in files
> 
> ```python
> PyodideSandbox(files=[
>   {
>      "bytes": ...,
>      "path": ...,
>   },
>   {
> 
>   }
> ])
> ```
> 
> Semantics are to create the directory on the pyodide side if it doesn't exist.
> 
> This also simplifies a lot of the interface since we don't need to worry about:
> 
>     1. creating directories
> 
>     2. get_attached_files
> 
>     3. clear_filesystem_operations
> 
>     4. as_structured_tool
> 
> 
> And in a separate PR we can worry about getting files **out** of the sandbox

Sorry for the delay in responding. I've had a chance to look at this now.

I really like your third approach and agree it's the best solution. I've actually been experimenting with something similar in my tests:

```python
sandbox = PyodideSandbox(files={
    "data.csv": "name,age\nJohn,25",
    "image.png": image_bytes,
    "models/config.json": json.dumps({"param": 0.5})
})
```

Regarding retrieving files from the sandbox after execution, I'm thinking about implementing a get_file method on the result object, something like:

```python
result = sandbox.execute(code)
output_csv = result.get_file('output.csv')
```

I'll tackle this as a separate PR. I'm not sure if I can complete the output feature this weekend, but I'll definitely start working on it.

I want to be clear that I'll be closing the files/description part of the sandbox this week by resolving both the args and description implementation along with the attachment method.

After that, I'll open another PR specifically for the output functionality.

In the future, I'd love to discuss other features for the lib. I really like this project and plan to use it in my work to gradually replace e2b as it matures.

fullzer4 committed:
fix: return file output limitation | remove local test functions

fullzer4 committed:
fixes

fullzer4 committed:
fixes

fullzer4 committed:
fix: improve filesystem operations and dynamic tool descriptions

fullzer4 committed:
Merge pull request #4 from fullzer4/fix/conventions-and-patterns

Fix/conventions and patterns

fullzer4 committed:
fix(pyodide): unrestrict file system access to ensure Pyodide can load dependencies

fullzer4 committed:
fix: configure Deno permissions for reliable Pyodide execution

fullzer4 committed:
fix: restore uv.lock

fullzer4 committed:
fix: restore uv.lock

fullzer4 committed:
fix: ensure custom descriptions are preserved in PyodideSandboxTool

fullzer4 commented:
@eyurtsev I've completed the filesystem implementation with stdin. It's working well and solves the CLI argument size limitations we discussed.

I chose a binary protocol over JSON stdin for file transfer, implemented with ReadableStream for efficient binary data handling, because it preserves binary data integrity and avoids "Maximum call stack size exceeded" errors that occur when sending large files as JSON through stdin. Would appreciate your thoughts on this approach.

I'm already working on the output functionality with a get_file method to extract files from the sandbox after execution. Will submit this in a separate PR once it's ready and tested.

Regarding the permissions flags, I had to make changes to fix critical errors like:

```bash
Error: Requires write access to "/home/fullzer4/path/node_modules/.deno/pyodide@0.27.5/node_modules/pyodide/micropip-0.8.0-py3-none-any.whl"
```

and Code

```bash
NotCapable: Requires read access to "/path/node_modules/.deno/pyodide@0.27.5/node_modules/pyodide/pyodide.asm.wasm"
```

The approach follows the principle of least privilege while ensuring everything works properly.

For the description part, I implemented a template-based system with _description_template and _build_description() that dynamically shows available files to the LLM:

```python
A secure Python code sandbox with filesystem support... ATTACHED FILES AVAILABLE: • data.csv • config.json These files are already loaded and ready to use...
```

The implementation now guarantees that when users provide custom descriptions, they are respected exactly as specified, while still supporting dynamic file information through an optional placeholder system.

I kept the filesystem methods (attach_file, create_directory, etc.) as they provide a clean, intuitive API. These methods make working with sandbox files much simpler than reimplementing these operations in user code.

I've also implemented the constructor-based file attachment you suggested, so users can now provide files directly when creating the sandbox tool:

```python
sandbox_tool = PyodideSandboxTool(
    allow_net=True,
    files={
        "sales.csv": sales_data
    }
)
```

I decided to maintain both approaches - constructor-based initialization and method-based file operations - because they serve different use cases. The constructor approach is perfect for setting up files at creation time in a clean, declarative style, while the method-based approach allows for more dynamic file manipulation during the tool's lifecycle. This flexibility gives users the best of both worlds without forcing them into a single pattern.

Let me know if you'd like me to adjust anything before merging.

fullzer4 commented:
Hi @eyurtsev, I hope you're doing well! I just wanted to gently check in on the filesystem support PR when you have a moment. I'm eager to hear your thoughts. Thank you so much!

eyurtsev reviewed (commented):
Left a few questions and I'd like to remove the mutability and and various help methods on the Python side.

I can take care of the changes on the python side if you'd like me to!

If you want to get the changes in more quickly, we can also break this PR into smaller ones (e.g., just tackling the typescript side first.)

fullzer4 commented on libs/sandbox-py/langchain_sandbox/pyodide.py:
```diff
@@ -508,35 +784,85 @@ class State(AgentState):
+    _sync_sandbox: SyncPyodideSandbox | None = PrivateAttr(default=None)
+    _custom_description: bool = PrivateAttr(default=False)
+
+    def model_post_init(self, /, __context) -> None:
```
I used `model_post_init` specifically to handle Pydantic's field initialization lifecycle. When trying to update the `description` field after `super().__init__()`, Pydantic had already frozen the field state, causing silent failures when `attach_file()` tried to update the description dynamically.

`model_post_init` ensures the description is properly built after all Pydantic initialization is complete, allowing proper field mutation for the dynamic file information feature.

fullzer4 commented on libs/pyodide-sandbox-js/main.ts:
```diff
@@ -24,11 +27,111 @@ import pyodide_js  # noqa
+            exists = os.path.exists(path)
+            return {"success": True, "exists": exists}
+            
+        elif operation == "remove":
```
You're absolutely right to question these. I need to be honest here - I accidentally committed the `copy` operation while testing for future `get_file` functionality, and forgot to remove the `remove` and `exists` operations from earlier development iterations.

Looking at the actual implementation:
- **remove/exists**: These operations are NOT used by any of the helper methods and serve no purpose in the current functionality
- **copy**: This was accidentally left in from testing future output functionality (extracting files FROM the sandbox after execution)

The current file attachment works entirely through the binary streaming protocol via stdin, so these filesystem operations are completely unnecessary. I can remove all three (`remove`, `exists`, `copy`) to simplify the interface.

fullzer4 commented on libs/sandbox-py/langchain_sandbox/pyodide.py:
```diff
@@ -157,46 +162,240 @@ def __init__(
+
+        # These read paths are the minimum required for Pyodide to function properly
+        # Without these, Pyodide cannot load its core WASM files and other resources
+        read_paths = [
```
Thanks again for the feedback. I'm learning a lot from your insights.

Regarding the JS module not being unregistered during Pyodide execution, I hadn't noticed this earlier. I'll revert it to the previous state. However, for some tests, I'll have to temporarily allow broader read and write permissions to work around the WASM module permission errors in Pyodide.

These changes will be addressed in a separate PR along with the "Unregister js module in pyodide #5" issue.

fullzer4 committed:
refactor: remove unnecessary filesystem operations (exists, remove, copy)

fullzer4 committed:
revert: restore previous permission model and add allow_read to tests

fullzer4 commented:
> Left a few questions and I'd like to remove the mutability and and various help methods on the Python side.
> 
> I can take care of the changes on the python side if you'd like me to!
> 
> If you want to get the changes in more quickly, we can also break this PR into smaller ones (e.g., just tackling the typescript side first.)

Thanks for offering to help! I'm happy to handle the remaining changes on the Python side myself - there's no rush on my end to get these changes in immediately. I've already addressed some of your review points with recent commits (specifically the unnecessary filesystem operations and permission model reversion).

If you prefer breaking this into smaller PRs as you suggested, I'm open to that approach too, but I don't want you to worry about making the changes yourself unless you specifically want to. I'm comfortable implementing the remaining feedback items based on your guidance.

fullzer4 commented:
Hi @eyurtsev,

Just checking in on the recent suggestions regarding the PR:

    Regarding the helper methods that add mutability: I mentioned that I kept them for dynamic use cases, but I'm happy to refactor to a constructor-only approach if you prefer.
    About using Pydantic’s model_post_init: It's being used to ensure proper field initialization and allow dynamic updates to the description post-init (needed for the attach files functionality).

I left a comment marked "Left a few questions..." and would appreciate it if you could review the points I raised when you have a chance. No rush—just a reminder so the PR doesn't get stale.

rayshen92 commented:
Hi @fullzer4,

Thank you very much for submitting this PR. I tried running the example locally but ran into the following error:

```text
Error during execution:
Traceback (most recent call last):
  File "/lib/python312.zip/_pyodide/_base.py", line 597, in eval_code_async
    await CodeRunner(
  File "/lib/python312.zip/_pyodide/_base.py", line 411, in run_async
    coroutine = eval(self.code, globals, locals)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<exec>", line 1, in <module>
FileNotFoundError: [Errno 44] No such file or directory: 'input.txt'
```

Here’s the code I used:

```python
files = {
    "input.txt": "Hello world!"
}

# Define sandbox tool with filesystem support
sandbox_tool = PyodideSandboxTool(
    files=files,
    allow_net=True,
)

# Create an agent with the sandbox tool
agent = create_react_agent("anthropic:claude-3-7-sonnet-latest", [sandbox_tool])

query = "display the content of input.txt"

async def run_agent(query: str):
    async for chunk in agent.astream({"messages": query}):
        print(chunk)

if __name__ == "__main__":
    asyncio.run(run_agent(query))
```

Could you please advise how to resolve this? Did I miss a configuration step, or is there something I need to adjust in the sandbox setup?

Thank you for your help!

fullzer4 commented:
Hi @rayshen92 ,

The error occurred because the system was referencing an older version of the pyodide-sandbox-js component (e.g., "jsr:@langchain/pyodide-sandbox@0.0.4"). By patching the PKG_NAME to point to the local main.ts file, we ensure that the updated, local version is used to test both components together.

For example, in our tests we apply the following patch:

```python
@pytest.fixture
def pyodide_package(monkeypatch: pytest.MonkeyPatch) -> None:
    """Patch PKG_NAME to point to a local deno typescript file."""
    if os.environ.get("RUN_INTEGRATION", "").lower() == "true":
        # Skip this test if running in integration mode
        return
    local_script = str(current_dir / "../../../pyodide-sandbox-js/main.ts")
    monkeypatch.setattr("langchain_sandbox.pyodide.PKG_NAME", local_script)
```

Without this patch, you would be using the old version. Alternatively, to work without changing the PKG_NAME in tests, a new version would need to be generated—for example:

```python
PKG_NAME = "jsr:@langchain/pyodide-sandbox@0.0.7"  # new version example
```

This ensures that the correct version is used, which is why the error occurred when referencing the outdated component.

Hope this clears up the confusion!

fullzer4 commented on libs/sandbox-py/langchain_sandbox/pyodide.py:
```diff
@@ -611,8 +931,106 @@ class PyodideSandboxToolInput(BaseModel):
+        ):
+            self.description = self._build_description()
+
+    def attach_file(
```
Hi @eyurtsev,

Following up on my previous message about the PR feedback. I wanted to let you know that the output files functionality is now complete.

I'd like to wrap up this PR so I can focus on implementing other library features and get this version released. I'm still waiting for your feedback on those specific points I mentioned earlier (the helper methods for mutability and the Pydantic model_post_init approach).

Once we resolve those questions, I think we'll be in good shape to merge this and move forward with the next set of features.

Thanks for your time!

fullzer4 committed:
fix: Immutability and remove helper methods

fullzer4 commented:
Hi @eyurtsev,

Quick update: I revisited the code in production and confirmed your earlier point—the mutability helpers weren’t adding real value. I’ve now removed them, made the models fully immutable, and shifted all directory/file handling into the constructor.

Let me know if there’s anything else you’d like tweaked; otherwise I think we’re ready to merge.

Thanks again for the guidance!

eyurtsev commented:
@fullzer4 awesome sorry for the delay in responding. I'll will like commandeer to take over the changes.

brisacoder commented:
This would be a welcomed feature. Very recently I switched back to running LLM generated Python code in containers for the very reason that this sandbox cannot read and write from file system.  Many of my Langgraph applications read CSV/XLS files, analyze data, save reports, amongst others that need FS access.

fullzer4 committed:
Merge branch 'langchain-ai:main' into main

fullzer4 committed:
feat: pyodide new lib

fullzer4 committed:
reverse: local version

sreeram004 commented:
Thanks @fullzer4 for this. This is a really important and must have feature to fully support the goal of the package - execution of LLM generated code safely..

@eyurtsev Do you think this can be prioritised? 

Thanks again

kersta-staden commented:
This feature would be really welcome! That would really change a lot!
